Solution

BipBobBoop opened the pull request:
[DEMO LINK](https://BipBobBoop.github.io/react_todo-app/)

Done in full accordance with https://todomvc.com/examples/vanillajs/, and all assignment items are completed.

Anastasiia-Svintsova commented on src/components/TodosFilter/TodoFilter.tsx:
```diff
@@ -0,0 +1,30 @@
+          <a
+            href={`#/${status.toLowerCase()}`}
+            className={classNames({ selected: selectedFilter === status })}
+            onClick={() => handleFilterClick(status)}
```
It's recommended to prevent the default action for the onClick event to avoid any unexpected page reloads when the link is clicked.

Anastasiia-Svintsova commented on src/components/Main/Main.tsx:
```diff
@@ -0,0 +1,42 @@
+
+    updateTodos(todos.map((todo: Todo) => ({
+      ...todo,
+      completed: isToggleAllChecked ? Status.Completed : '',
```
The assignment of the 'completed' property is not correct. It should be a boolean value, not a string from the enum Status. If 'isToggleAllChecked' is true, then all todos should be marked as not completed (false), and vice versa.

Anastasiia-Svintsova commented on src/components/Main/Main.tsx:
```diff
@@ -0,0 +1,42 @@
+  const { todos, updateTodos } = useContext(TodosContext);
+  const [isToggleAllChecked, setIsToggleAllChecked] = useState(true);
+
+  const isAllChecked = useMemo(() => {
```
The condition inside the 'every' method should directly check if the 'completed' property of the todo is true, instead of comparing it with Status.Completed.

Anastasiia-Svintsova commented on src/components/Main/Main.tsx:
```diff
@@ -0,0 +1,42 @@
+
+export const Main = () => {
+  const { todos, updateTodos } = useContext(TodosContext);
+  const [isToggleAllChecked, setIsToggleAllChecked] = useState(true);
```
The initial value of 'isToggleAllChecked' should be false, as initially, not all todos are completed.

Anastasiia-Svintsova commented on src/components/NewTodo/NewTodo.tsx:
```diff
@@ -0,0 +1,62 @@
+  };
+
+  const addNewTodo = () => {
+    const newTodo: Todo = {
```
The 'id' of the newTodo is being set as 'todos.length + 1'. If a todo is deleted, the length will decrease and you may end up with duplicate ids for different todos. Consider using a unique id generator for the 'id' field.
![image](https://github.com/mate-academy/react_todo-app/assets/99403198/16615bb6-e959-4746-825e-42bbaf506e88)

Anastasiia-Svintsova commented on src/components/Footer/Footer.tsx:
```diff
@@ -0,0 +1,43 @@
+  const { todos, updateTodos } = useContext(TodosContext);
+
+  const handleClearCompleted = () => {
+    updateTodos(todos.filter(todo => todo.completed !== 'completed'));
```
The completed property of a todo item is likely a boolean so checking for !== 'completed' seems incorrect. Instead, it should be checking for !todo.completed

Anastasiia-Svintsova commented on src/components/TodoList/TodoList.tsx:
```diff
@@ -0,0 +1,27 @@
+
+  return (
+    <ul className="todo-list" data-cy="todoList">
+      {filteredTodos && filteredTodos.map(
```
Here, `filteredTodos &&` is not necessary because `filteredTodos` is never `undefined` or `null`. It's always an array, even if it's empty, so you can safely map over it.

Anastasiia-Svintsova commented on src/components/TodoList/TodoList.tsx:
```diff
@@ -0,0 +1,27 @@
+export const TodoList = () => {
+  const { todos, filter } = useContext(TodosContext);
+
+  let filteredTodos = todos;
```
Instead of initializing `filteredTodos = todos`, you can define filteredTodos inside a useEffect or useMemo hook which will recompute todos when `todos` or `filter` changes. This will optimize your component to avoid unnecessary computations on each render.

Anastasiia-Svintsova commented on src/components/TodoItem/TodoItem.tsx:
```diff
@@ -0,0 +1,143 @@
+  const [value, setValue] = useState(item.title);
+  const [completed, setCompleted] = useState(item.completed);
+  const inputRef = useRef<HTMLInputElement | null>(null);
+  const hasEditingTodo = todos.some(todo => todo.completed === 'editing');
```
The `completed` property of a todo is used to check if a todo is being edited. This is confusing because `completed` should indicate if a todo is completed or not. Consider using a different property to show if a todo item is being edited.

Anastasiia-Svintsova commented on src/components/Footer/Footer.tsx:
```diff
@@ -0,0 +1,43 @@
+    todo => todo.completed === 'completed',
+  ).length;
+
+  if (todos.length === 0) {
```
Use logical operators to simplify your code 
```suggestion
  if (!todos.length) {
```

Anastasiia-Svintsova commented on src/components/Footer/Footer.tsx:
```diff
@@ -0,0 +1,43 @@
+    updateTodos(todos.filter(todo => todo.completed !== 'completed'));
+  };
+
+  const amountOfUncompletedTodos = todos.filter(
```
Use more clear naming for variables. It's hard to read this name for now. activeTodos or activeTodosCount maybe?

Anastasiia-Svintsova commented on src/components/NewTodo/NewTodo.tsx:
```diff
@@ -0,0 +1,62 @@
+    const newTodo: Todo = {
+      id: todos.length + 1,
+      title: value,
+      completed: '',
```
The 'completed' field in the 'newTodo' object should be a boolean, not a string. Consider setting the initial value to false.
![image](https://github.com/mate-academy/react_todo-app/assets/99403198/09cbc9e1-a44c-467b-a91c-f5ebfeb7a660)

Anastasiia-Svintsova commented on src/components/TodoItem/TodoItem.tsx:
```diff
@@ -0,0 +1,143 @@
+  item: Todo,
+};
+
+/* eslint-disable jsx-a11y/control-has-associated-label */
```
Don't disable eslint. Try to fix errors instead

Anastasiia-Svintsova commented on src/components/TodoItem/TodoItem.tsx:
```diff
@@ -0,0 +1,143 @@
+          onChange={() => handleToggleItem(item.id)}
+        />
+        <label {...labelProps}>
+          {item.title}
```
Use destructuring for objects

Anastasiia-Svintsova reviewed (changes requested):
Great start! Check my comments, please

BipBobBoop commented:
@Anastasiia-Svintsova Thank you for your valuable feedback! I've implemented the changes you suggested and refactored the code accordingly. Please take a look and let me know if everything looks good now. Your insights were truly helpful!

BipBobBoop commented:
> Great start! Check my comments, please

@Anastasiia-Svintsova Thank you for your valuable feedback! I've implemented the changes you suggested and refactored the code accordingly. Please take a look and let me know if everything looks good now. Your insights were truly helpful!

Anastasiia-Svintsova reviewed (changes requested):
You've done a good job! One comment is left. Fix it, please 
- Check styles for the completed todos 
![image](https://github.com/mate-academy/react_todo-app/assets/99403198/68b7a32f-a8d3-4c4b-9a70-aae823aeefbe)

BipBobBoop commented:
> You've done a good job! One comment is left. Fix it, please
> 
> * Check styles for the completed todos
>   ![image](https://user-images.githubusercontent.com/99403198/262671170-68b7a32f-a8d3-4c4b-9a70-aae823aeefbe.png)

![image](https://github.com/mate-academy/react_todo-app/assets/95710670/929e35e6-9315-4629-be3a-844739c83092)
![image](https://github.com/mate-academy/react_todo-app/assets/95710670/770b1a9b-bf1b-40e3-86ba-f808f7bad220)

maxim2310 reviewed (approved):
Well done!!!
